### PR TITLE
Integer unification part 1 : move Fixnum/Bignum methods to Integer

### PR DESCRIPTION
**Pull request by eregon**
* To prepare or help the transition to Ruby 2.4.
* All methods are moved from Fixnum and Bignum to Integer, but they are still separate classes, which is very compatible.
* The bigger change to define `Fixnum=Bignum=Integer` is separate and will be done with the upgrade to 2.4 as it requires a lot more changes and needs MRI 2.4 tests and 2.4 specs.

### CODE REVIEW

File: spec/truffle/methods_spec.rb [hunk 1]
```diff
@@ -8,17 +8,18 @@
 
 require_relative '../ruby/spec_helper'
 
-# Switch to MRI 2.3.3 and run:
+# Switch to MRI 2.3.7 and run:
```
Comment (chrisseaton) [inline spec/truffle/methods_spec.rb]: Maybe say 'compatible version of MRI' so we don't have to update in the future?

File: src/main/java/org/truffleruby/core/numeric/IntegerNodes.java [hunk 2]
```diff
@@ -469,9 +575,42 @@ public Object mod(long a, DynamicObject b) {
             return mod;
         }
 
+        @TruffleBoundary
+        @Specialization
+        public Object mod(DynamicObject a, long b) {
+            if (b == 0) {
+                throw new ArithmeticException("divide by zero");
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/core/numeric/IntegerNodes.java]: I know you just moved this code, but shouldn't this be a Ruby exception?

File: src/main/java/org/truffleruby/core/numeric/IntegerNodes.java [hunk 3]
```diff
@@ -469,9 +575,42 @@ public Object mod(long a, DynamicObject b) {
...
+            return fixnumOrBignum(Layouts.BIGNUM.getValue(a).mod(BigInteger.valueOf(b)));
+        }
+
+        @TruffleBoundary
+        @Specialization(guards = "isRubyBignum(b)")
+        public Object mod(DynamicObject a, DynamicObject b) {
+            final BigInteger bigint = Layouts.BIGNUM.getValue(b);
+            final int compare = bigint.compareTo(BigInteger.ZERO);
+            if (compare == 0) {
+                throw new ArithmeticException("divide by zero");
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/core/numeric/IntegerNodes.java]: Same question here about the exception type.

File: src/main/java/org/truffleruby/core/numeric/IntegerNodes.java [hunk 4]
```diff
@@ -888,6 +888,30 @@ public int compare(long a, DynamicObject b) {
...
+
+        @Specialization(guards = "!isInfinity(b)")
+        public int compare(DynamicObject a, double b) {
+            return Double.compare(Layouts.BIGNUM.getValue(a).doubleValue(), b);
+        }
+
+        @Specialization(guards = "isInfinity(b)")
+        public int compareInfinity(DynamicObject a, double b) {
+            if (b < 0) {
+                return +1;
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/core/numeric/IntegerNodes.java]: I don't think we do this explicit `+1` in other compare method implementations.

File: src/main/java/org/truffleruby/core/numeric/IntegerNodes.java [hunk 5]
```diff
@@ -587,6 +555,22 @@ public Object mod(DynamicObject a, long b) {
             return fixnumOrBignum(Layouts.BIGNUM.getValue(a).mod(BigInteger.valueOf(b)));
         }
 
+        @TruffleBoundary
+        @Specialization
+        public double mod(DynamicObject a, double b) {
+            if (b == 0) {
+                throw new ArithmeticException("divide by zero");
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/core/numeric/IntegerNodes.java]: Same question here about the exception type.

File: src/main/ruby/core/array.rb [hunk 6]
```diff
@@ -475,7 +475,8 @@ def fetch(idx, default=undefined)
       end
     end
 
-    if left >= Fixnum::MAX || right > Fixnum::MAX
+    unless Truffle.invoke_primitive(:fixnum_fits_into_long, left) &&
```
Comment (nirvdrum) [inline src/main/ruby/core/array.rb]: Was it significant that left was a `>=` check while right was a `>` check?

File: src/main/ruby/core/range.rb [hunk 7]
```diff
@@ -187,7 +187,7 @@ def hash
     hash ^= self.end.hash << 9
     hash ^= excl << 24;
     # Are we throwing away too much here for a good hash value distribution?
-    hash & Fixnum::MAX
+    hash & 0x7fffffffffffffff
```
Comment (nirvdrum) [inline src/main/ruby/core/range.rb]: I don't like the magic value here, but this whole method has magic values so maybe it doesn't matter.

File: src/main/java/org/truffleruby/core/numeric/IntegerNodes.java (hunk 2 above)
Comment (eregon) [reply src/main/java/org/truffleruby/core/numeric/IntegerNodes.java]: It's translated by ExceptionTranslationNode.

File: src/main/java/org/truffleruby/core/numeric/IntegerNodes.java (hunk 4 above)
Comment (eregon) [reply src/main/java/org/truffleruby/core/numeric/IntegerNodes.java]: It's a nice symmetry, but I can remove if you prefer.

File: src/main/ruby/core/array.rb (hunk 6 above)
Comment (eregon) [reply src/main/ruby/core/array.rb]: Doesn't look like so, I think it was a bug.

File: src/main/ruby/core/range.rb (hunk 7 above)
Comment (eregon) [reply src/main/ruby/core/range.rb]: We could have a `LONG_MAX` constant somewhere, but where?

File: src/main/java/org/truffleruby/core/numeric/IntegerNodes.java (hunk 4 above)
Comment (nirvdrum) [reply src/main/java/org/truffleruby/core/numeric/IntegerNodes.java]: I don't feel strongly enough about it. It looks slightly odd to me, but if you like it, feel free to keep it.

File: src/main/ruby/core/range.rb (hunk 7 above)
Comment (nirvdrum) [reply src/main/ruby/core/range.rb]: In the `Truffle` module maybe? Maybe a comment here is good enough.

File: src/main/ruby/core/array.rb (hunk 6 above)
Comment (eregon) [reply src/main/ruby/core/array.rb]: Actually, it looks like `left` is always checked by `coerce_to_collection_length`, which already checks that it is a Fixnum.
Same for `right`, so I should be able to remove that check completely.

Comment (eregon) [reply src/main/ruby/core/array.rb]: I misread, those values are still modified after `coerce_to_collection_length`.
So I think it was indeed a bug to use `>=`.

File: src/main/ruby/core/range.rb (hunk 7 above)
Comment (eregon) [reply src/main/ruby/core/range.rb]: Added under `Truffle::Platform`.

**Pull request merged**